### PR TITLE
Feature/develop create recipe flow

### LINKED ISSUE #32: Develop "Search recipe" flow

### LINKED ISSUE #50: Named and optional argument usage;

### LINKED ISSUE #53: Regex usage;

### LINKED ISSUE #54: Widening and narrowing type conversions;

### LINKED ISSUE #55: LINQ to Objects usage (methods and queries);

### PR DESCRIPTION
**Pull request by onlyidev**
- Full CRUD for recipes
- Pagination
- Search recipes by name (closes #32)
- Minor migration updates

**Completed requirements**:
- closes #50
- closes #53 
- closes #55 
- closes #54

### CODE REVIEW

File: Controllers/RecipeController.cs [hunk 1]
```diff
@@ -17,27 +20,75 @@ public RecipeController(AppDatabaseContext db)
     }
 
     [HttpGet]
-    public IActionResult getRecipies()
+    [Authorize]
+    public IActionResult getRecipies(uint page = 1, string? filter = null) // Query parameter with default value
```
Comment (onlyidev) [inline Controllers/RecipeController.cs:24]: Closes #50

File: Controllers/RecipeController.cs [hunk 2]
```diff
@@ -17,27 +20,75 @@ public RecipeController(AppDatabaseContext db)
...
+            return BadRequest();
+        }
+
+        if (filter != null)
+        {
+            string str = filter.map(letter => $".*?{Regex.Escape(letter.ToString().ToLower())}.*?");
+            return Ok(_db.Recipes.Where(c => Regex.IsMatch(c.Name.ToLower(), str)).Take(AppDatabaseContext.ItemsPerPage));
+        }
+
+        return Ok(_db.Recipes.Skip(((int)(page) - 1) * AppDatabaseContext.ItemsPerPage).Take(AppDatabaseContext.ItemsPerPage));
```
Comment (onlyidev) [inline Controllers/RecipeController.cs:37]: Closes #54 (widening part as `page` is of type `uint`)

Comment (onlyidev) [inline Controllers/RecipeController.cs:37]: Closes #55

File: Controllers/RecipeController.cs (hunk 1 above)
Comment (onlyidev) [inline Controllers/RecipeController.cs:24]: Optional and named argument usage

Comment (onlyidev) [reply Controllers/RecipeController.cs:24]: Closes #50

**Comment by onlyidev** [comment]
Closes:
- #50

File: Controllers/RecipeController.cs [hunk 3]
```diff
@@ -17,27 +20,75 @@ public RecipeController(AppDatabaseContext db)
...
-        // TODO implement filter functionality (will use regex)
-        return Ok("getRecipies");
+        if (page < 1)
+        {
+            return BadRequest();
+        }
+
+        if (filter != null)
+        {
+            string str = filter.map(letter => $".*?{Regex.Escape(letter.ToString().ToLower())}.*?");
```
Comment (onlyidev) [inline Controllers/RecipeController.cs:33]: Regex usage

File: Controllers/RecipeController.cs (hunk 2 above)
Comment (onlyidev) [inline Controllers/RecipeController.cs:37]: LINQ usage

Comment (onlyidev) [inline Controllers/RecipeController.cs:37]: Type conversions widening part

File: Controllers/RecipeController.cs [hunk 4]
```diff
@@ -17,27 +20,75 @@ public RecipeController(AppDatabaseContext db)
...
+    public IActionResult deleteRecipe([FromBody] dynamic req) // We only need the id of the recipe
     {
-        // TODO remove from database
-        return Ok("deleteRecipe");
+        RecipeModel? recipe = null;
+        dynamic data = JsonConvert.DeserializeObject<dynamic>(req.ToString());
+        try
+        {
+            if (data.id != null
+                && (recipe = _db.Recipes.Find((int)data.id.Value)) != null
```
Comment (onlyidev) [inline Controllers/RecipeController.cs:77]: Type conversions narrowing part

**Review by onlyidev — COMMENTED** [review]
Requirement implementations

File: Controllers/RecipeController.cs (hunk 1 above)
Comment (Storas008) [inline Controllers/RecipeController.cs:24]: must use PascalCase, please change to GetRecipies();

File: Controllers/RecipeController.cs [hunk 5]
```diff
@@ -17,27 +20,75 @@ public RecipeController(AppDatabaseContext db)
...
+            return Ok(_db.Recipes.Where(c => Regex.IsMatch(c.Name.ToLower(), str)).Take(AppDatabaseContext.ItemsPerPage));
+        }
+
+        return Ok(_db.Recipes.Skip(((int)(page) - 1) * AppDatabaseContext.ItemsPerPage).Take(AppDatabaseContext.ItemsPerPage));
     }
 
     [HttpPost]
     [Authorize]
-    public IActionResult upsertRecipe(RecipeModel recipe)
+    public IActionResult upsertRecipe(RecipeModel recipe) //For upserting we need the full model information (id can be ommited for creating a new recipe)
```
Comment (Storas008) [inline Controllers/RecipeController.cs:42]: Please change to UpsertRecipe();

File: Controllers/RecipeController.cs [hunk 6]
```diff
@@ -17,27 +20,75 @@ public RecipeController(AppDatabaseContext db)
...
+        {
+            return BadRequest();
+        }
+
     }
 
     [HttpDelete]
     [Authorize]
-    public IActionResult deleteRecipe(RecipeModel recipe)
+    public IActionResult deleteRecipe([FromBody] dynamic req) // We only need the id of the recipe
```
Comment (Storas008) [inline Controllers/RecipeController.cs:70]: change to PascalCase

**Review by Storas008 — CHANGES_REQUESTED** [review]
Everything seems right (lol), just need some very minor changes.

**Review by Storas008 — APPROVED** [review]

**Pull request merged**